torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -259,12 +263,31 @@ void ProcessGroupAgent::sync() {
 }
 
 void ProcessGroupAgent::start() {
+  start_.store(true);
   listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
 }
 
+void ProcessGroupAgent::localShutdown() {
+  if (!start_.load()) {
+    return;
+  }
+  LOG(INFO) << "Stopping ProcessGroupAgent.";
+  // CV to make sure we don't abort it in an invalid state.
+  {
+    std::unique_lock<std::mutex> lock(recvWorkMutex_);
+    recvWorkCV_.wait(lock, [this]() { return recvWork_; });
+    recvWork_->abort();
+    start_.store(false);
```

Comment:
Shouldn't this be set at the beginning?

Reply:
If this is set at the beginning, it is possible that the listenerThread would not have started, in which case, `recvWork` would never be valid and we'd wait on the above condition forever. Since we wait on the condition, we'd let the listener thread run at least once and put `recvWork` in a valid state, and then we'd abort it, and prevent the listener thread from running any further

Reply:
I don't think we should rely on recvWork being in a valid state during shutdown.